src/mlpack/tests/main_tests/hmm_training_tests.cpp

```diff
@@ -0,0 +1,209 @@
+
+using namespace mlpack;
+
+struct HMMTrainMainTestFixture
+{
+ public:
+  HMMTrainMainTestFixture()
+  {
+    // Cache in the options for this program.
+    CLI::RestoreSettings(testName);
+  }
+
+  ~HMMTrainMainTestFixture()
+  {
+    // Clear the settings.
+    bindings::tests::CleanMemory();
+    CLI::ClearSettings();
+  }
+};
+
+BOOST_FIXTURE_TEST_SUITE(HMMTrainMainTest, HMMTrainMainTestFixture);
+
+inline void fileExists(std::string fileName)
+{
+  ifstream ifp(fileName);
+  if (!ifp.good())
+    BOOST_FAIL("Bad stream " + fileName);
+  ifp.close();
+}
+
+// Make sure that the number of states cannot be negative
+BOOST_AUTO_TEST_CASE(HMMTrainStatesTest)
+{
+  std::string inputFileName = "hmm_train_obs.csv";
+  int states = -3;  // Invalid!
+  std::string hmmType = "discrete";
+
+  fileExists(inputFileName);
+  SetInputParam("input_file", std::move(inputFileName));
+  SetInputParam("states", states);
+  SetInputParam("type", std::move(hmmType));
+
+  Log::Fatal.ignoreInput = true;
+  BOOST_REQUIRE_THROW(mlpackMain(), std::runtime_error);
+  Log::Fatal.ignoreInput = false;
+}
+
+// Make sure that tolerance is non negative
+BOOST_AUTO_TEST_CASE(HMMTrainToleranceNonNegative)
+{
+  std::string inputFileName = "hmm_train_obs.csv";
+  int states = 3;
+  std::string hmmType = "gaussian";
+  double tol = - 100;  // Invalid
+
+  fileExists(inputFileName);
+  SetInputParam("input_file", std::move(inputFileName));
+  SetInputParam("states", states);
+  SetInputParam("type", std::move(hmmType));
+  SetInputParam("tolerance", tol);
+
+  Log::Fatal.ignoreInput = true;
+  BOOST_REQUIRE_THROW(mlpackMain(), std::runtime_error);
+  Log::Fatal.ignoreInput = false;
+}
+
+// Make sure an error is thrown if type is something other than
+// "discrete", "gaussian" or "gmm"
+BOOST_AUTO_TEST_CASE(HMMTrainTypeTest)
+{
+  std::string inputFileName = "hmm_train_obs.csv";
+  int states = 3;
+  std::string hmmType = "some-not-supported-possibly-non-type";
+
+  fileExists(inputFileName);
+  SetInputParam("input_file", std::move(inputFileName));
+  SetInputParam("states", states);
+  SetInputParam("type", std::move(hmmType));
+
+  Log::Fatal.ignoreInput = true;
+  BOOST_REQUIRE_THROW(mlpackMain(), std::runtime_error);
+  Log::Fatal.ignoreInput = false;
+}
+
+// Make sure that the number of gaussians cannot be less than 0
+BOOST_AUTO_TEST_CASE(HMMTrainGaussianTest)
+{
+  std::string inputFileName = "hmm_train_obs.csv";
+  int states = 3;
+  std::string hmmType = "gmm";
+  int gaussians = -2;
+
+  fileExists(inputFileName);
+  SetInputParam("input_file", std::move(inputFileName));
+  SetInputParam("states", states);
+  SetInputParam("type", std::move(hmmType));
+  SetInputParam("gaussians", gaussians);
+
+  Log::Fatal.ignoreInput = true;
+  BOOST_REQUIRE_THROW(mlpackMain(), std::runtime_error);
+  Log::Fatal.ignoreInput = false;
+}
+
+// Make sure that model reuse is possible and work properly
+BOOST_AUTO_TEST_CASE(HMMTrainReuseModelTest)
```

Comment:
I think it might be useful to also test this when training without labels (so the Baum-Welch algorithm is used for training instead of direct estimation).